### PR TITLE
Add learning rate scheduler

### PR DESCRIPTION
**Pull request by lumurillo**
# What does this PR do?

PR moved from internal MART repo.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by mzweilin — CHANGES_REQUESTED** [review]
Could you add some test cases for the feature?

File: mart/attack/perturber/batch.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 __all__ = ["BatchPerturber"]
 
 
-class BatchPerturber(Callback, torch.nn.Module):
+class BatchPerturber(torch.nn.Module):
```
Comment (lumurillo) [inline mart/attack/perturber/batch.py:16]: Since the `Callback` parent class is removed, the `on_run_start` method here is still needed?

File: mart/attack/adversary.py [hunk 2]
```diff
@@ -128,21 +128,25 @@ def done(self) -> bool:
...
         # Set up the optimizer.
         self.cur_iter = 0
 
         # We could be at the inference/no-grad mode here.
         # Initialize lazy module
         # FIXME: Perturbers can just use on_run_start/on_run_end to initialize
-        self.perturber(input, target)
+        # self.perturber(input, target)
+        # Initialize the Uninitialized parameters here, because self.opt needs perturber.params.
+        adversary.perturber.on_run_start(adversary, input, target, model, **kwargs)
```
[the remark this answers is not in the document]
Comment (lumurillo) [reply mart/attack/adversary.py:140]: this change is provoking that one of the Adversary's unit tests fail.

File: mart/attack/adversary.py [hunk 3]
```diff
@@ -128,21 +128,25 @@ def done(self) -> bool:
...
-        super().on_run_start(adversary, input, target, model, **kwargs)
-
         # FIXME: We should probably just register IterativeAdversary as a callback.
         # Set up the optimizer.
         self.cur_iter = 0
 
         # We could be at the inference/no-grad mode here.
         # Initialize lazy module
         # FIXME: Perturbers can just use on_run_start/on_run_end to initialize
-        self.perturber(input, target)
```
[the remark this answers is not in the document]
Comment (lumurillo) [reply mart/attack/adversary.py:140]: I think that `LazyModuleMixin` already does it when calling the **forward** method: https://github.com/pytorch/pytorch/blob/1a0738f5992807475aeb47bd93b847c7bdc356f0/torch/nn/modules/lazy.py#L179

File: mart/attack/perturber/batch.py (hunk 1 above)
Comment (lumurillo) [reply mart/attack/perturber/batch.py:16]: If this is not going to be a Callback, you shouldn't add the perturbation in the callbacks list in the Adversary's `IterativeGenerator`: https://github.com/IntelLabs/MART/blob/4bed8926fb91ee0b16d798936f0db8bfe8092190/mart/attack/adversary.py#L101-L103
